File: testing/unit-tests/ruby/RubyPokemonTest/Gen3PokemonTest.rb
#!/usr/bin/ruby
#
# Copyright (c) 2017 Nicholas Corgan (n.corgan@gmail.com)
#
# Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
# or copy at http://opensource.org/licenses/MIT)
#

require_relative "PokemonTest"
require_relative "PokemonTestParams"

require "PKMN"

require "minitest/autorun"
require "os"
require "pathname"

class Gen3PokemonTest < PokemonTest
    @@CONTEST_TYPES = ["Cool", "Beauty", "Cute", "Smart", "Tough"]
    @@CONTEST_LEVELS = ["Super", "Hyper", "Master"]
    @@RIBBONS = ["Champion", "Winning", "Victory", "Artist",
                 "Effort", "Marine", "Land", "Sky",
                 "Country", "National", "Earth", "World"]

    @@RNG = Random.new

    def _check_initial_ribbons_map(pokemon)
        @@CONTEST_TYPES.each do |contest_type|
            assert(pokemon.ribbons.keys.include?(contest_type))

            @@CONTEST_LEVELS.each do |contest_level|
                ribbon_name = contest_type + " " + contest_level
                assert(pokemon.ribbons.keys.include?(ribbon_name))
            end
        end

        @@RIBBONS.each do |ribbon|
            assert(pokemon.ribbons.keys.include?(ribbon))
        end
    end

    def _test_contest_ribbons(pokemon)
        @@CONTEST_TYPES.each do |contest_type|
            ribbon_name = contest_type
            super_ribbon_name = contest_type + " Super"
            hyper_ribbon_name = contest_type + " Hyper"
            master_ribbon_name = contest_type + " Master"

            pokemon.ribbons[hyper_ribbon_name] = true
            assert(pokemon.ribbons[ribbon_name])
            assert(pokemon.ribbons[super_ribbon_name])
            assert(pokemon.ribbons[hyper_ribbon_name])
            assert(!pokemon.ribbons[master_ribbon_name])

            pokemon.ribbons[super_ribbon_name] = false
            assert(pokemon.ribbons[ribbon_name])
            assert(!pokemon.ribbons[super_ribbon_name])
            assert(!pokemon.ribbons[hyper_ribbon_name])
            assert(!pokemon.ribbons[master_ribbon_name])
        end
    end

    def _test_ribbons(pokemon)
        @@RIBBONS.each do |ribbon|
            random_bool = (@@RNG.rand(100) > 50)
            pokemon.ribbons[ribbon] = random_bool
            assert_equal(random_bool, pokemon.ribbons[ribbon])
        end
    end

    def _gen3_test_common(pokemon)
        # Gender and personality are tied, so make sure they affect each other.
        pokemon.gender = PKMN::Gender::FEMALE
        assert_operator(pokemon.personality & 0xFF, :<, 0xFF)
        pokemon.gender = PKMN::Gender::MALE

        pokemon.personality = 0x1234AB00
        assert_equal(PKMN::Gender::FEMALE, pokemon.gender)
        pokemon.personality = 0xCD5678FF
        assert_equal(PKMN::Gender::MALE, pokemon.gender)

        # Setting shininess should affect personality.
        pokemon.is_shiny = false
        assert(!pokemon.is_shiny)
        pokemon.is_shiny = true
        assert(pokemon.is_shiny)

        _check_initial_ribbons_map(pokemon)
        _test_contest_ribbons(pokemon)
        _test_ribbons(pokemon)
    end

    def _gba_pokemon_test(game_params)
        pokemon = PKMN::Pokemon.new(game_params[1], game_params[0], "", 30)

        test_params = PokemonTestParams.new
        test_params.valid_ball = "Great Ball"
        test_params.invalid_balls = ["Friend Ball", "Heal Ball"]
        test_params.valid_item = "Razz Berry"
        test_params.invalid_items = ["Berry", "Mach Bike"]
        test_params.expected_original_location = "Fateful encounter"
        test_params.valid_locations = ["Petalburg Woods", "Viridian Forest"]
        test_params.invalid_locations = ["New Bark Town", "Twinleaf Town"]
        test_params.valid_moves = ["Swallow", "Flamethrower", "Return", "Fire Blast"]
        test_params.invalid_moves = ["Shadow Sky", "Roost"]
        test_params.valid_original_games = ["Ruby", "Sapphire", "Emerald", "FireRed",
                                            "LeafGreen", "Colosseum", "XD", "Colosseum/XD"]
        test_params.invalid_original_games = ["Gold", "HeartGold"]

        pokemon_test_common(pokemon, test_params)
        _gen3_test_common(pokemon)
    end

    def _gcn_pokemon_test(game_params)
        pokemon = PKMN::Pokemon.new(game_params[1], game_params[0], "", 30)

        test_params = PokemonTestParams.new
        test_params.valid_ball = "Great Ball"
        test_params.invalid_balls = ["Friend Ball", "Heal Ball"]
        test_params.valid_item = "Razz Berry"
        test_params.invalid_items = ["Berry", "Mach Bike"]
        test_params.expected_original_location = "Distant land"
        test_params.valid_locations = ["Phenac City", "Orre Colosseum"]
        test_params.invalid_locations = ["New Bark Town", "Twinleaf Town"]
        test_params.valid_moves = ["Swallow", "Flamethrower", "Return", "Fire Blast"]
        test_params.invalid_moves = ["Roost", "Flame Burst"]
        test_params.valid_original_games = ["Ruby", "Sapphire", "Emerald", "FireRed",
                                            "LeafGreen", "Colosseum", "XD", "Colosseum/XD"]
        test_params.invalid_original_games = ["Gold", "HeartGold"]

        pokemon_test_common(pokemon, test_params)
        _gen3_test_common(pokemon)
    end

    [["Ruby", "Torchic"],
     ["Sapphire", "Mudkip"],
     ["Emerald", "Treecko"],
     ["FireRed", "Charmander"],
     ["LeafGreen", "Bulbasaur"]].each do |game_params|
        define_method("test_gba_pokemon_#{game_params[0]}") do
            _gba_pokemon_test(game_params)
        end
    end

    [["Colosseum", "Espeon"],
     ["XD", "Umbreon"]].each do |game_params|
        define_method("test_gcn_pokemon_#{game_params[0]}") do
            _gcn_pokemon_test(game_params)
        end
    end
end

File: testing/unit-tests/c/c_pokemon_test/gen3_pokemon_test.c
/*
 * Copyright (c) 2017-2018 Nicholas Corgan (n.corgan@gmail.com)
 *
 * Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
 * or copy at http://opensource.org/licenses/MIT)
 */

#include "c_test_common.h"

#include "pokemon_test.h"

#include <pkmntest-c/util.h>

#include <pkmn-c.h>

#include <string.h>

#define STRBUFFER_LEN 1024
static char strbuffer[STRBUFFER_LEN];
static enum pkmn_error error = PKMN_ERROR_NONE;

static const char* contest_types[] =
{
    "Cool", "Beauty", "Cute", "Smart", "Tough"
};

static const char* contest_levels[] =
{
    "", " Super", " Hyper", " Master"
};

static const char* ribbons[] =
{
    "Champion", "Winning", "Victory", "Artist",
    "Effort", "Marine", "Land", "Sky",
    "Country", "National", "Earth", "World"
};

static void check_initial_ribbons(
    struct pkmn_pokemon* p_pokemon
)
{
    TEST_ASSERT_NOT_NULL(p_pokemon);

    struct pkmn_string_list ribbon_names;
    error = pkmn_pokemon_get_ribbon_names(
                p_pokemon,
                &ribbon_names
            );
    PKMN_TEST_ASSERT_SUCCESS(error);
    TEST_ASSERT_EQUAL(32, ribbon_names.length);

    for(int type_index = 0; type_index < 5; ++type_index)
    {
        for(int level_index = 0; level_index < 4; ++level_index)
        {
            snprintf(
                strbuffer,
                sizeof(strbuffer),
                "%s%s",
                contest_types[type_index],
                contest_levels[level_index]
            );

            TEST_ASSERT_TRUE(
                string_list_contains(&ribbon_names, strbuffer)
            );
        }
    }

    for(int i = 0; i < 12; ++i)
    {
        TEST_ASSERT_TRUE(
            string_list_contains(&ribbon_names, ribbons[i])
        );
    }

    error = pkmn_string_list_free(&ribbon_names);
    PKMN_TEST_ASSERT_SUCCESS(error);
}

static void test_contest_ribbons(
    struct pkmn_pokemon* p_pokemon
)
{
    TEST_ASSERT_NOT_NULL(p_pokemon);

    char super_ribbon_name[STRBUFFER_LEN] = {0};
    char hyper_ribbon_name[STRBUFFER_LEN] = {0};
    char master_ribbon_name[STRBUFFER_LEN] = {0};
    bool has_ribbon = false;

    for(int type_index = 0; type_index < 5; ++type_index)
    {
        snprintf(
            super_ribbon_name,
            sizeof(super_ribbon_name),
            "%s Super",
            contest_types[type_index]
        );
        snprintf(
            hyper_ribbon_name,
            sizeof(hyper_ribbon_name),
            "%s Hyper",
            contest_types[type_index]
        );
        snprintf(
            master_ribbon_name,
            sizeof(master_ribbon_name),
            "%s Master",
            contest_types[type_index]
        );

        error = pkmn_pokemon_set_has_ribbon(
                    p_pokemon,
                    hyper_ribbon_name,
                    true
                );
        PKMN_TEST_ASSERT_SUCCESS(error);

        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    contest_types[type_index],
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_TRUE(has_ribbon);
        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    super_ribbon_name,
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_TRUE(has_ribbon);
        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    hyper_ribbon_name,
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_TRUE(has_ribbon);
        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    master_ribbon_name,
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_FALSE(has_ribbon);
    }
}

static void test_ribbons(
    struct pkmn_pokemon* p_pokemon
)
{
    TEST_ASSERT_NOT_NULL(p_pokemon);

    bool has_ribbon = false;
    for(int ribbon_index = 0; ribbon_index < 12; ++ribbon_index)
    {
        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    ribbons[ribbon_index],
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_FALSE(has_ribbon);

        error = pkmn_pokemon_set_has_ribbon(
                    p_pokemon,
                    ribbons[ribbon_index],
                    true
                );
        PKMN_TEST_ASSERT_SUCCESS(error);

        error = pkmn_pokemon_has_ribbon(
                    p_pokemon,
                    ribbons[ribbon_index],
                    &has_ribbon
                );
        PKMN_TEST_ASSERT_SUCCESS(error);
        TEST_ASSERT_TRUE(has_ribbon);
    }
}

static void gba_pokemon_test(
    enum pkmn_species species,
    enum pkmn_game game
)
{
    enum pkmn_error error = PKMN_ERROR_NONE;
    struct pkmn_pokemon pokemon =
    {
        .species = PKMN_SPECIES_NONE,
        .game = PKMN_GAME_NONE,
        .p_internal = NULL
    };

    error = pkmn_pokemon_init(
                species,
                game,
                "",
                30,
                &pokemon
            );
    PKMN_TEST_ASSERT_SUCCESS(error);
    TEST_ASSERT_NOT_NULL(pokemon.p_internal);

    pkmn_test_values_t test_values =
    {
        .valid_ball = PKMN_BALL_GREAT_BALL,
        .p_invalid_balls = (enum pkmn_ball[]){
            PKMN_BALL_FRIEND_BALL,
            PKMN_BALL_HEAL_BALL,
            PKMN_BALL_NONE
        },

        .valid_item = PKMN_ITEM_RAZZ_BERRY,
        .p_invalid_items = (enum pkmn_item[]){
            PKMN_ITEM_BERRY,
            PKMN_ITEM_MACH_BIKE,
            PKMN_ITEM_NONE
        },

        .expected_original_location = "Fateful encounter",
        .valid_locations = (char*[]){"Petalburg Woods", "Viridian Forest", NULL},
        .invalid_locations = (char*[]){"New Bark Town", "Twinleaf Town", NULL},

        .moves = (char*[]){"Swallow", "Flamethrower", "Return", "Fire Blast", NULL},
        .invalid_moves = (char*[]){"Shadow Sky", "Roost", NULL},

        .valid_original_games = (enum pkmn_game[]){
            PKMN_GAME_RUBY, PKMN_GAME_SAPPHIRE, PKMN_GAME_EMERALD,
            PKMN_GAME_FIRERED, PKMN_GAME_LEAFGREEN,
            PKMN_GAME_COLOSSEUM, PKMN_GAME_XD,
            PKMN_GAME_NONE
        },
        .invalid_original_games = (enum pkmn_game[]){
            PKMN_GAME_GOLD, PKMN_GAME_HEARTGOLD,
            PKMN_GAME_NONE
        }
    };

    pokemon_test_common(
        &pokemon,
        &test_values
    );

    // Test ribbons.
    check_initial_ribbons(&pokemon);
    test_contest_ribbons(&pokemon);
    test_ribbons(&pokemon);

    error = pkmn_pokemon_free(&pokemon);
    PKMN_TEST_ASSERT_SUCCESS(error);
}

static void gcn_pokemon_test(
    enum pkmn_species species,
    enum pkmn_game game
)
{
    enum pkmn_error error = PKMN_ERROR_NONE;
    struct pkmn_pokemon pokemon =
    {
        .species = PKMN_SPECIES_NONE,
        .game = PKMN_GAME_NONE,
        .p_internal = NULL
    };

    error = pkmn_pokemon_init(
                species,
                game,
                "",
                30,
                &pokemon
            );
    PKMN_TEST_ASSERT_SUCCESS(error);
    TEST_ASSERT_NOT_NULL(pokemon.p_internal);

    pkmn_test_values_t test_values =
    {
        .valid_ball = PKMN_BALL_GREAT_BALL,
        .p_invalid_balls = (enum pkmn_ball[]){
            PKMN_BALL_FRIEND_BALL,
            PKMN_BALL_HEAL_BALL,
            PKMN_BALL_NONE
        },

        .valid_item = PKMN_ITEM_RAZZ_BERRY,
        .p_invalid_items = (enum pkmn_item[]){
            PKMN_ITEM_BERRY,
            PKMN_ITEM_MACH_BIKE,
            PKMN_ITEM_NONE
        },

        .expected_original_location = "Distant land",
        .valid_locations = (char*[]){"Phenac City", "Orre Colosseum", NULL},
        .invalid_locations = (char*[]){"New Bark Town", "Twinleaf Town", NULL},

        .moves = (char*[]){"Swallow", "Flamethrower", "Return", "Fire Blast", NULL},
        .invalid_moves = (char*[]){"Roost", "Flame Burst", NULL},

        .valid_original_games = (enum pkmn_game[]){
            PKMN_GAME_RUBY, PKMN_GAME_SAPPHIRE, PKMN_GAME_EMERALD,
            PKMN_GAME_FIRERED, PKMN_GAME_LEAFGREEN,
            PKMN_GAME_COLOSSEUM, PKMN_GAME_XD,
            PKMN_GAME_NONE
        },
        .invalid_original_games = (enum pkmn_game[]){
            PKMN_GAME_GOLD, PKMN_GAME_HEARTGOLD,
            PKMN_GAME_NONE
        }
    };

    pokemon_test_common(
        &pokemon,
        &test_values
    );

    // Test ribbons.
    check_initial_ribbons(&pokemon);
    test_contest_ribbons(&pokemon);
    test_ribbons(&pokemon);

    error = pkmn_pokemon_free(&pokemon);
    PKMN_TEST_ASSERT_SUCCESS(error);
}

void ruby_pokemon_test()
{
    gba_pokemon_test(PKMN_SPECIES_TORCHIC, PKMN_GAME_RUBY);
}

void sapphire_pokemon_test()
{
    gba_pokemon_test(PKMN_SPECIES_MUDKIP, PKMN_GAME_SAPPHIRE);
}

void emerald_pokemon_test()
{
    gba_pokemon_test(PKMN_SPECIES_TREECKO, PKMN_GAME_EMERALD);
}

void firered_pokemon_test()
{
    gba_pokemon_test(PKMN_SPECIES_CHARMANDER, PKMN_GAME_FIRERED);
}

void leafgreen_pokemon_test()
{
    gba_pokemon_test(PKMN_SPECIES_BULBASAUR, PKMN_GAME_LEAFGREEN);
}

void colosseum_pokemon_test()
{
    gcn_pokemon_test(PKMN_SPECIES_ESPEON, PKMN_GAME_COLOSSEUM);
}

void xd_pokemon_test()
{
    gcn_pokemon_test(PKMN_SPECIES_UMBREON, PKMN_GAME_XD);
}
